src/induction/base.ml

```diff
@@ -193,18 +221,40 @@ let rec next (input_sys, aparam, trans, solver, k, unknowns) =
     (* Asserting implications if k > 0. *)
     if Numeral.(k > zero) then
       nu_unknowns
+      (* If we're skipping steps, we don't need to assert that the property
+         held in the previous step *)
+      |> List.filter (fun (name, _) -> match TransSys.get_prop_kind trans name with
+        | Reachable Some (From b)
+        | Reachable Some (At b) -> b < k_int
+        | Reachable Some (FromWithin (b, _))-> b < k_int
+        | _ -> true
+      ) 
       |> List.map (fun (_, term) -> Term.bump_state Numeral.(k-one) term)
       |> Term.mk_and |> SMTSolver.assert_term solver ;
 
+    (* (Temporarily) filter out properties with a lower bound smaller than
+       current k. *)
+    let out_of_bounds, nu_unknowns = nu_unknowns |> List.partition (
+      fun (name,_) -> match TransSys.get_prop_kind trans name with
+      | Reachable Some (From b)
+      | Reachable Some (At b) -> b > k_int
+      | Reachable Some (FromWithin (b, _))-> b > k_int
+      | _ -> false
+    ) in
     (* Filtering properties which are not known to be k-true at this step. *)
     let unknowns_at_k, k_true =
       nu_unknowns |> List.partition (
-        fun (name,_) -> match TransSys.get_prop_status trans name with
-        | Property.PropKTrue k -> k_int > k
+        fun (name,_) -> match (TransSys.get_prop_status trans name, TransSys.get_prop_kind trans name) with
+        | (Property.PropKTrue k, Invariant) 
+        | (Property.PropKTrue k, Reachable None) -> k_int > k
+        | (Property.PropKTrue k, Reachable Some (From b)) -> k_int > k && b <= k_int
+        | (Property.PropKTrue k, Reachable Some (Within b)) -> k_int > k && b >= k_int
+        | (Property.PropKTrue k, Reachable Some (At b)) -> k_int > k && b = k_int
+        | (Property.PropKTrue k, Reachable Some (FromWithin (b1, b2))) -> k_int > k && b1 <= k_int && b2 >= k_int
```

Comment:
Kind 2 should do three things. (1) Classifying all the (negated) reachability properties with a lower bound `L` as K-true for `L-1` from the beginning, (2) classifying all the reachability properties with an upper bound `U` as unreachable as soon as they are K-true for `U`, and  (3) filtering out all the reachability properties that are out of bounds. This code only addresses the latter (and with some issues). 

If you address the first point, then you don't need to perform the first partition between `out_of_bounds` and `nu_unknowns` because the partition between `unknowns_at_k` and `k_true` would take care of reachability properties with a lower bound. Moreover, the second partition would be as simple as before (`| Property.PropKTrue k -> k_int > k`) because it would depend only on the property being K-true.

If you address the second point too, then Kind 2 will not include rechability properties out of the upper bound either (because they are already proven).